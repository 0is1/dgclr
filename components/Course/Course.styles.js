import styled from 'styled-components';

const Styles = {
  Wrapper: styled.div`
    margin: 1rem;
    padding: 2rem;
  `,
  Title: styled.h2`
    font-size: 3rem;
    font-weight: 800;
  `,
  Description: styled.p`
    font-weight: 200;
    line-height: 1.25;
  `,
};

export default Styles;
