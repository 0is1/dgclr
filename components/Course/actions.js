// @flow
import type { Course } from 'lib/types';

export const actionTypes = {
  SET_COURSES: 'dgclr/SET_COURSES',
};

/**
 * Set courses action
 * @param {Array} courses
 * @return {Object} action of SET_COURSES
 */
export const setCourses = (courses: Array<Course> = []) => ({
  type: actionTypes.SET_COURSES,
  courses,
});
