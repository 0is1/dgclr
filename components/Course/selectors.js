// @flow
import { get } from 'lodash/fp';
import type { State } from 'lib/types';

export const courseBySlugFromState = (state: State, ownProps: { slug: string }) => get(['courses', ownProps.slug], state) || {};

export default courseBySlugFromState;
