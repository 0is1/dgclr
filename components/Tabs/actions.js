// @flow
export const actionTypes = {
  SET_TABS: 'DGCLR_SET_TABS',
};

/**
 * Set tabbar data
 * @param {String} id of tabs
 * @param {Int} activeIndex of tabs
 * @return {Object} action of SET_TABS
 */
export const setTabs = (id: string, activeIndex: number) => ({
  type: actionTypes.SET_TABS,
  id,
  activeIndex,
});
