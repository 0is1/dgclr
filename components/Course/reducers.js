import { keyBy } from 'lodash/fp';
import update from 'updeep';
import { createReducer } from 'lib/createReducer';
import { actionTypes } from './actions';

export const initialState = {};

const reducer = createReducer({
  [actionTypes.SET_COURSES]: (state, action) => {
    const courses = keyBy(course => course.slug, action.courses);
    return update(courses, state);
  },
});

export default reducer;
