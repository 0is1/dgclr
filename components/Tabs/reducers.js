import u from 'updeep';
import { createReducer } from 'lib/createReducer';
import { actionTypes } from './actions';

export const initialState = {
  data: {},
};

const reducer = createReducer({
  [actionTypes.SET_TABS]: (state, action) => {
    const { id, activeIndex } = action;
    return u.updateIn(['data', id], { id, activeIndex }, state);
  },
});

export default reducer;
