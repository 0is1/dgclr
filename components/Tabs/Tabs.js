// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Tabs } from 'rebass';
import { getActiveIndex } from 'components/Tabs/selectors';
import { setTabs as setTabsFunc } from 'components/Tabs/actions';
import type { State as ReduxState } from 'lib/types';
import Tab from './Tab';

type Props = {
  id: string,
  tabs: [string],
};
type MapStateToProps = { activeIndex: ?number };
type MapDispatchToProps = { setTabs: Function };

type CombinedProps = Props & MapStateToProps & MapDispatchToProps;

class TabsComponent extends Component<CombinedProps> {
  componentDidMount() {
    const { activeIndex, id, setTabs } = this.props;
    if (!activeIndex) {
      setTabs(id, 0);
    }
  }

  render() {
    const { activeIndex, id, tabs } = this.props;
    if (typeof activeIndex !== 'number') return null;
    const props = { activeIndex, id };
    const tabData = tabs.map((name, index) => {
      const key = `${name}-${index}`;
      return <Tab {...props} key={key} index={index} name={name} />;
    });
    return <Tabs>{tabData}</Tabs>;
  }
}

const mapStateToProps = (state: ReduxState, ownProps: Props): MapStateToProps => ({
  activeIndex: getActiveIndex(state, ownProps),
});

const mapDispatchToProps = (dispatch: Function): MapDispatchToProps => ({
  setTabs: (id, activeIndex = 0) => dispatch(setTabsFunc(id, activeIndex)),
});

export default connect<CombinedProps, Props, MapStateToProps, any, any, Function>(
  mapStateToProps,
  mapDispatchToProps,
)(TabsComponent);
