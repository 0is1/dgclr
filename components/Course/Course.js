// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { size } from 'lodash';
import { Helmet } from 'react-helmet';
import ReactHtmlParser from 'react-html-parser';
import { Card, Flex, Subhead } from 'rebass';
import {
  convertCoordinatesToObject,
  convertLinksToHtml,
  courseAddressDetails,
  getCourseMapUrlForLayout,
  getTitle,
  uniqueLayoutRatings,
} from 'helpers/utils';
import { courseBySlugFromState } from 'components/Course/selectors';
import { setCourses as setCoursesFunc } from 'components/SearchContainer/actions';
import Map from 'components/Map';
import Layouts from 'components/Layouts';
import Tabs from 'components/Tabs';
import LayoutRatingBadges from 'components/Layout/Badges';
import { ClipLoader } from 'components/Spinners';
import Image from 'components/Image';
import { getActiveIndex } from 'components/Tabs/selectors';
import Styles from 'components/Course/Course.styles';
import BaseStyles from 'components/Container/Container.styles';
import type { Course as CourseType, GraphQLData, State as ReduxState } from 'lib/types';

type Props = {
  data: GraphQLData,
};
type OwnProps = Props & { slug: string };
type MapStateToProps = { activeIndex: ?number, course: CourseType };
type MapDispatchToProps = { setCourses: Function };
type CombinedProps = Props & MapStateToProps & MapDispatchToProps;

const {
  Box, Description, PanelWrapper, PanelHeader, PanelFooter, Strong, Title,
} = Styles;

const { BaseText } = BaseStyles;

class Course extends Component<CombinedProps> {
  componentDidMount() {
    const { course, data = {} } = this.props;
    const { courseBySlug } = data;
    if (size(course) < 1 && courseBySlug && courseBySlug.length) {
      this.setCourses(courseBySlug);
    }
  }

  componentDidUpdate(prevProps: CombinedProps, prevState: any, snapshot: any) {
    if (snapshot !== null) {
      const { course, data } = this.props;
      const { courseBySlug } = data;
      if (size(course) < 1 && courseBySlug && courseBySlug.length) {
        this.setCourses(courseBySlug);
      }
    }
  }

  getSnapshotBeforeUpdate(prevProps: CombinedProps) {
    const { data = {} } = this.props;
    const { courseBySlug } = data;
    if (size(prevProps.course) < 1 && courseBySlug && courseBySlug.length) {
      return true;
    }
    return null;
  }

  setCourses = (courses: Array<CourseType>) => {
    const { setCourses } = this.props;
    setCourses(courses);
  };

  render() {
    const { activeIndex, course, data = {} } = this.props;
    const { courseBySlug = [] } = data;
    if ((size(course) < 1 && size(courseBySlug) < 1) || (data && data.loading)) {
      return <ClipLoader />;
    }

    const courseData = size(course) > 0 ? course : courseBySlug[0];
    const {
      name, courseInfo, description, locationInfo, layouts,
    } = courseData;
    const mapUrl = getCourseMapUrlForLayout(course.layouts, activeIndex || 0);
    const descriptionWithLinks = convertLinksToHtml(description);
    const { location } = locationInfo;
    const coordinates = location && convertCoordinatesToObject(location.coordinates);
    const layoutNames = layouts.map(layout => layout.name);
    const layoutTabs = <Tabs tabs={layoutNames} id={courseData._id} />;
    const mapElement = coordinates ? <Map coordinates={coordinates} data={courseData} /> : null;
    const courseImage = mapUrl && mapUrl !== '' ? <Image alt={name} src={mapUrl} /> : null;
    const ratings = uniqueLayoutRatings(layouts);
    return (
      <Box width={[1, 1, 1, '90%']} p={[0, 0, '0 0.5rem', '0 1rem']} m={['.5rem 0', '.5rem 0', '.5rem 0', '1rem auto']}>
        <Helmet>
          <title>{getTitle(name)}</title>
        </Helmet>
        <PanelWrapper>
          <PanelHeader>Radan tiedot</PanelHeader>
          <Flex flexWrap="wrap">
            <Box width={[1, 1, 1, 1 / 2]} p={[2, 2, null, 4]}>
              <Title>
                {name}
                <LayoutRatingBadges ratings={ratings} />
              </Title>
              <Description>{ReactHtmlParser(descriptionWithLinks)}</Description>
              <PanelWrapper mt={4}>
                <PanelHeader>Radan lis√§tiedot:</PanelHeader>
                <Box width="100%" px={1} pt={2} pb="1rem">
                  {courseInfo.founded && (
                    <BaseText>
                      <Strong>Perustettu: </Strong>
                      {`${courseInfo.founded}`}
                    </BaseText>
                  )}
                  {courseInfo.courseDesigner && (
                    <BaseText>
                      <Strong>Suunnittelija: </Strong>
                      {`${courseInfo.courseDesigner}`}
                    </BaseText>
                  )}
                  {courseInfo.rangeMaster && (
                    <BaseText>
                      <Strong>Ratamestari: </Strong>
                      {`${courseInfo.rangeMaster}`}
                    </BaseText>
                  )}
                  {courseInfo.basketType && (
                    <BaseText>
                      <Strong>Korityyppi: </Strong>
                      {`${courseInfo.basketType}`}
                    </BaseText>
                  )}
                  {courseInfo.infoSignType && (
                    <BaseText>
                      <Strong>Opastaulut: </Strong>
                      {`${courseInfo.infoSignType}`}
                    </BaseText>
                  )}
                  {courseInfo.teeType && (
                    <BaseText>
                      <Strong>Heittoalustat: </Strong>
                      {`${courseInfo.teeType}`}
                    </BaseText>
                  )}
                  {courseInfo.courseTypes && (
                    <BaseText>
                      <Strong>Ratatyypit: </Strong>
                      {`${courseInfo.courseTypes.join(', ')}`}
                    </BaseText>
                  )}
                  {courseInfo.surfaceShapeTypes && (
                    <BaseText>
                      <Strong>Pinnanmuodot: </Strong>
                      {`${courseInfo.surfaceShapeTypes.join('. ')}`}
                    </BaseText>
                  )}
                  {courseInfo.fee && courseInfo.fee.value && (
                    <BaseText>
                      <Strong>Maksullinen / Ilmainen: </Strong>
                      {`${courseInfo.fee.value}`}
                    </BaseText>
                  )}
                </Box>
              </PanelWrapper>
            </Box>
            <Box width={[1, 1, 1, 1 / 2]} p={[2, 2, null, 4]}>
              <Card width="100%">
                {mapElement}
                <Subhead mt={2} p={2}>
                  {courseAddressDetails(locationInfo)}
                </Subhead>
              </Card>
            </Box>
          </Flex>
          <PanelFooter>
            <Flex flexWrap="wrap">
              <Box width={[1, 1, 1, 1 / 2]} p={2}>
                {layoutTabs}
                <Layouts id={course._id} layouts={layouts} />
              </Box>
              <Box width={[1, 1, 1, 1 / 2]} p={2}>
                {courseImage}
              </Box>
            </Flex>
          </PanelFooter>
        </PanelWrapper>
      </Box>
    );
  }
}

const mapStateToProps = (state: ReduxState, ownProps: OwnProps): MapStateToProps => {
  const course = courseBySlugFromState(state, ownProps);
  const id = (course && course._id) || false;
  return {
    activeIndex: getActiveIndex(state, { id }),
    course,
  };
};

const mapDispatchToProps = (dispatch: Function): MapDispatchToProps => ({
  setCourses: courses => dispatch(setCoursesFunc(courses)),
});

export default connect<CombinedProps, OwnProps, MapStateToProps, any, any, Function>(
  mapStateToProps,
  mapDispatchToProps,
)(Course);
