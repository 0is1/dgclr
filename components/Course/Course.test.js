// Link.react.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Card } from 'rebass';
import TestStoreProvider from 'jest/TestStoreProvider';
import mockCoursesData from 'components/SearchContainer/mock/courses.mock';
import Tabs from 'components/Tabs/Tabs';
import { DEFAULT_STATE } from 'lib/rootReducer';
import Styles from './Course.styles';
import Course from './Course';

const { Description, PanelWrapper } = Styles;

const defaultProps = {
  setCourses: jest.fn(),
  slug: mockCoursesData[1].slug,
};

function setup() {
  const storeData = {
    ...DEFAULT_STATE,
    search: {
      ...DEFAULT_STATE.search,
      courses: {
        [mockCoursesData[0].slug]: mockCoursesData[0],
        [mockCoursesData[1].slug]: mockCoursesData[1],
      },
    },
  };
  const components = (
    <TestStoreProvider storeData={storeData}>
      <Course {...defaultProps} />
    </TestStoreProvider>
  );
  const enzymeWrapper = mount(components);
  return { enzymeWrapper };
}

describe('Course component', () => {
  it('Match snapshot', () => {
    const component = renderer.create(
      <TestStoreProvider storeData={DEFAULT_STATE}>
        <Course {...defaultProps} />
      </TestStoreProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Renders correctly', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper).toBeDefined();
    expect(enzymeWrapper.find(PanelWrapper).length).toEqual(2);
    expect(enzymeWrapper.find(Card).length).toEqual(1);
    expect(enzymeWrapper.props().children.props.slug).toEqual(mockCoursesData[1].slug);
    expect(enzymeWrapper.find('h2').length).toEqual(1);
    expect(enzymeWrapper.find('h2').text()).toEqual('Ford DiscGolfParkAA1');
    expect(enzymeWrapper.find(Description).length).toEqual(1);
    expect(enzymeWrapper.find(Description).text()).toEqual(mockCoursesData[1].description);
    expect(enzymeWrapper.find(Tabs).length).toEqual(1);
  });
});
