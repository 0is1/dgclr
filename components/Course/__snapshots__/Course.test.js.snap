// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course component Match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(withI18nextTranslation(Course))
    setCourses={[MockFunction]}
    slug="ford-discgolfpark-tuusula"
    t={[MockFunction]}
  />
</Provider>
`;
